ksp_plugin/orbit_analyser.cpp

```diff
@@ -0,0 +1,212 @@
+#include "physics/discrete_trajectory.hpp"
+
+namespace principia {
+namespace ksp_plugin {
+namespace internal_orbit_analyser {
+
+using geometry::Frame;
+using physics::DiscreteTrajectory;
+using physics::MasslessBody;
+using physics::BodyCentredNonRotatingDynamicFrame;
+using quantities::IsFinite;
+
+OrbitAnalyser::OrbitAnalyser(not_null<Ephemeris<Barycentric>*> const ephemeris,
+                             Ephemeris<Barycentric>::FixedStepParameters const
+                                 analysed_trajectory_parameters)
+    : ephemeris_(ephemeris),
+      analysed_trajectory_parameters_(analysed_trajectory_parameters),
+      analyser_([this] { RepeatedlyAnalyseOrbit(); }) {}
+
+OrbitAnalyser::~OrbitAnalyser() {
+  if (analyser_.joinable()) {
+    keep_analysing_ = false;
+    analyser_.join();
+  }
+}
+
+void OrbitAnalyser::RequestAnalysis(
+    Instant const& first_time,
+    DegreesOfFreedom<Barycentric> const& first_degrees_of_freedom,
+    Time const& mission_duration,
+    not_null<RotatingBody<Barycentric> const*> primary) {
+  Ephemeris<Barycentric>::Guard guard(ephemeris_);
+  if (ephemeris_->t_min() > first_time) {
+    // Too much has been forgotten; we cannot perform this analysis.
+    return;
+  }
+  absl::MutexLock l(&lock_);
+  parameters_ = {std::move(guard),
+                 first_time,
+                 first_degrees_of_freedom,
+                 mission_duration,
+                 primary};
+}
+
+void OrbitAnalyser::RefreshAnalysis() {
+  absl::MutexLock l(&lock_);
+  if (next_analysis_.has_value()) {
+    analysis_ = next_analysis_;
+    next_analysis_.reset();
+  }
+}
+
+OrbitAnalyser::Analysis* OrbitAnalyser::analysis() {
+  return analysis_.has_value() ? &*analysis_ : nullptr;
+}
+
+int8_t OrbitAnalyser::next_analysis_percentage() const {
+  return next_analysis_percentage_;
+}
+
+void OrbitAnalyser::RepeatedlyAnalyseOrbit() {
+  while (keep_analysing_) {
+    // No point in going faster than 50 Hz.
+    std::chrono::steady_clock::time_point const wakeup_time =
+        std::chrono::steady_clock::now() + std::chrono::milliseconds(20);
+
+    std::optional<Parameters> parameters;
+    {
+      absl::ReaderMutexLock l(&lock_);
+      if (!parameters_.has_value()) {
+        // No parameters, let's wait for them to appear.
+        continue;
+      }
+      std::swap(parameters, parameters_);
+    }
+
+    Analysis analysis{parameters->first_time,
+                      parameters->primary};
+    DiscreteTrajectory<Barycentric> trajectory;
+    trajectory.Append(parameters->first_time, parameters->first_degrees_of_freedom);
+    std::vector<not_null<DiscreteTrajectory<Barycentric>*>> trajectories = {
+        &trajectory};
+    auto instance = ephemeris_->NewInstance(
+        trajectories,
+        Ephemeris<Barycentric>::NoIntrinsicAccelerations,
+        analysed_trajectory_parameters_);
+    for (Instant t =
+             parameters->first_time + parameters->mission_duration / 100;
```

Comment:
Why do we have this number, 100?  Is it just for the percentage?  How do we know that this yields a reasonable step?

Reply:
Now it's `0x1p10`. Note that this is unrelated to the integration step, there is an underlying fixed-step instance; that number is just the number of opportunities for interruption (and the granularity of the progress bar).